### DISCUSSION TITLE
Installation, Windows, GPU

### DISCUSSION THREAD

**Discussion by akraskovSMDB** [neuropoly/axondeepseg#752] [Help needed]
I have to reinstall axondeepseg on my Windows GPU computer (within miniconda). (I had a working installation before but I updated various individual packages to deal with "security issues" as requested by our glorious IT and I broke it somehow).

The problem is that when I follow instructions from the axondeepseg website I do not get GPU support
```
axondeepseg_test
Computing the segmentation of axon and myelin on test image.
2023-08-07 16:10:20.949 | INFO     | ivadomed.utils:define_device:160 - Cuda is not available.
2023-08-07 16:10:20.952 | INFO     | ivadomed.utils:define_device:161 - Working on cpu.  
```
I had the same problem previously and managed to solve it by installing cuda support separately from pytorch.
But now I can't, and it is on exactly the same computer. Essentially conda is not able to resolve conflicts. 

I still have a yml file exported with conda from the working installation, but it does not work now because of the conflict 

The conflict is caused by:
```
    The user requested torch==1.11.0
    ivadomed 2.9.8 depends on torch<=1.11.0 and >=1.8.1
    torchaudio 0.13.1 depends on torch==1.13.1
To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict
```
I tried 2, removed the specific versions for torch and torchaudio, installation worked.
Here are the differences in yml files (torch is the same but torchaudio is downgraded to 0.11.0)
```
sdiff -s -i -E -Z -b -a ads_venv_git_pytorch_loose.yml ads_venv_git_pytorch_new.yml
  - docutils=0.18.1=py38haa244fe_1                            <
  - pytorch=1.13.1=py3.8_cuda11.6_cudnn8_0                    <
  - sphinx=6.1.3=pyhd8ed1ab_0                                 <
                                                              >       - cachey==0.2.1
                                                              >       - docstring-parser==0.15
                                                              >       - docutils==0.17.1
                                                              >       - freetype-py==2.4.0
                                                              >       - heapdict==1.0.1
                                                              >       - hsluv==5.0.3
                                                              >       - in-n-out==0.1.8
                                                              >       - lazy-loader==0.3
                                                              >       - mypy-extensions==1.0.0
                                                              >       - napari-plugin-engine==0.2.0
                                                              >       - pint==0.21.1
                                                              >       - psygnal==0.9.1
                                                              >       - pydantic==1.10.12
                                                              >       - pyproject-hooks==1.0.0
                                                              >       - sphinx==4.5.0
                                                              >       - superqt==0.5.0
      - torch                                                 |       - tomli-w==1.0.0
      - torchaudio                                            |       - torch==1.11.0
                                                              >       - torchaudio==0.11.0
                                                              >       - vispy==0.12.2
```
BUT there is no CUDA support!
```
>python -m torch.utils.collect_env
Collecting environment information...
PyTorch version: 1.11.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Enterprise
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.16 | packaged by conda-forge | (default, Feb  1 2023, 15:53:35) [MSC v.1929 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: False
CUDA runtime version: 11.6.124
GPU models and configuration: GPU 0: Quadro RTX 5000
Nvidia driver version: 512.15
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] numpydoc==1.5.0
[pip3] torch==1.11.0
[pip3] torchaudio==0.11.0
[pip3] torchio==0.18.88
[pip3] torchvision==0.12.0
[conda] blas                      1.0                         mkl    conda-forge
[conda] mkl                       2021.4.0           haa95532_640
[conda] mkl-service               2.4.0            py38h2bbff1b_0
[conda] mkl_fft                   1.3.1            py38h277e83a_0
[conda] mkl_random                1.2.2            py38hf11a4ad_0
[conda] mypy-extensions           1.0.0                    pypi_0    pypi
[conda] numpy                     1.23.5           py38h3b20f71_0
[conda] numpy-base                1.23.5           py38h4da318b_0
[conda] numpydoc                  1.5.0                    pypi_0    pypi
[conda] pytorch-cuda              11.6                 h867d48c_1    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch                     1.11.0                   pypi_0    pypi
[conda] torchaudio                0.11.0                   pypi_0    pypi
[conda] torchio                   0.18.88                  pypi_0    pypi
[conda] torchvision               0.12.0                   pypi_0    pypi
```
 
As a sanity check, I can install pytorch in a separate environment and cuda is working there.
```
conda install pytorch==1.13.1 torchvision==0.14.1 torchaudio==0.13.1 pytorch-cuda=11.6 -c pytorch -c nvidia
```
```
(pytorchcuda) C:\miniconda\axondeepseg>python -m torch.utils.collect_env
Collecting environment information...
PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Enterprise
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.17 (default, Jul  5 2023, 20:44:21) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: True
CUDA runtime version: 11.6.124
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Quadro RTX 5000
Nvidia driver version: 512.15
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==1.13.1
[pip3] torchaudio==0.13.1
[pip3] torchvision==0.14.1
[conda] blas                      1.0                         mkl
[conda] mkl                       2023.1.0         h8bd8f75_46356
[conda] mkl-service               2.4.0            py38h2bbff1b_1
[conda] mkl_fft                   1.3.6            py38hf11a4ad_1
[conda] mkl_random                1.2.2            py38hf11a4ad_1
[conda] numpy                     1.24.3           py38h79a8e48_1
[conda] numpy-base                1.24.3           py38h8a87ada_1
[conda] pytorch                   1.13.1          py3.8_cuda11.6_cudnn8_0    pytorch
[conda] pytorch-cuda              11.6                 h867d48c_1    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.13.1                   pypi_0    pypi
[conda] torchvision               0.14.1                   pypi_0    pypi

```

Could you please guide me, I am not sure what else I can try

Thank you

**Comment by mathieuboudreau** [comment]
If you're trying to install via your "saved" conda env, it may cause some issues as when it saves it, it sets all values to required versions.

eg,

`torchaudio==0.11.0`

is not a requirement in our native venv files, so when you try to install that with a newer / looser IVADOMED/torch version, it will cause issues. If you're set on trying it this way, I'd say remove the line I quoted above from your venv, and it should natively install a compatible version (it's not a required version for either ADS or IVADOMED's venv files).

If you're having more issues installing the GPU version on Windows, maybe someone in the IVADOMED github repo would know some potential solutions (I only work on Mac, so unfortunately I can't really test this). See here: https://github.com/ivadomed/ivadomed

Lastly, if you're not training a new model, it shouldn't cause any major issue for you to use the CPU version, beyond a bit longer processing time.

**Reply by akraskovSMDB** [reply]
Hi, does the default ads installation/environment suppose to add support for gpu/cuda?

I am doing all these bugivugi with previous export yml and exact versions just because for me the default one does not work.

Yes, I do need train models.

**Reply by Stoyan-I-A** [reply]
> does the default ads installation/environment suppose to add support for gpu/cuda?

The default ADS installation does not specify if Ivadomed should use CPU or GPU. It is handeled on the Ivadomed side.
You could try to [install Ivadomed](https://ivadomed.org/installation.html#step-2-install-ivadomed) on its own (independent from ADS) and see if it allows you to use your GPU.

**Reply by akraskovSMDB** [reply]
standalone ivadomed installation still does not add GPU/CUDA support 
Instead, I tried fist to install pytorch in a fresh python=3.8 environment 
```
conda install pytorch==1.11.0 torchvision==0.12.0 torchaudio==0.11.0 cudatoolkit=11.3 -c pytorch
```
(had to downgrade to cudatoolkit 11.3 because ivadomed wants torchaudio==0.11.0)
managed to install ivadomed with GPU/CUDA support 

BUT now ADS installation fails 
<details>
  <summary>Output of ADS installation showing error</summary>
<p>
Collecting pyzmq>=20 (from ipykernel>=5.2.0->napari-console>=0.0.6->napari->napari-ADS==1.0.0)
  Downloading pyzmq-25.1.1.tar.gz (1.4 MB)
     ---------------------------------------- 1.4/1.4 MB 9.6 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [25 lines of output]
      Compiling zmq\backend\cython\_device.pyx because it depends on C:\Users\nak179\AppData\Local\Temp\pip-build-env-27nrbhku\overlay\Lib\site-packages\Cython\Includes\libc\string.pxd.
      Compiling zmq\backend\cython\_poll.pyx because it depends on C:\Users\nak179\AppData\Local\Temp\pip-build-env-27nrbhku\overlay\Lib\site-packages\Cython\Includes\libc\string.pxd.
      Compiling zmq\backend\cython\_proxy_steerable.pyx because it depends on C:\Users\nak179\AppData\Local\Temp\pip-build-env-27nrbhku\overlay\Lib\site-packages\Cython\Includes\libc\string.pxd.
      Compiling zmq\backend\cython\context.pyx because it depends on C:\Users\nak179\AppData\Local\Temp\pip-build-env-27nrbhku\overlay\Lib\site-packages\Cython\Includes\libc\string.pxd.
      Compiling zmq\backend\cython\error.pyx because it depends on C:\Users\nak179\AppData\Local\Temp\pip-build-env-27nrbhku\overlay\Lib\site-packages\Cython\Includes\cpython\version.pxd.
      Compiling zmq\backend\cython\message.pyx because it depends on C:\Users\nak179\AppData\Local\Temp\pip-build-env-27nrbhku\overlay\Lib\site-packages\Cython\Includes\libc\string.pxd.
      Compiling zmq\backend\cython\socket.pyx because it depends on C:\Users\nak179\AppData\Local\Temp\pip-build-env-27nrbhku\overlay\Lib\site-packages\Cython\Includes\libc\string.pxd.
      Compiling zmq\devices\monitoredqueue.pyx because it depends on C:\Users\nak179\AppData\Local\Temp\pip-build-env-27nrbhku\overlay\Lib\site-packages\Cython\Includes\libc\string.pxd.
      [1/8] Cythonizing zmq\backend\cython\error.pyx
      [2/8] Cythonizing zmq\backend\cython\message.pyx
      [3/8] Cythonizing zmq\backend\cython\socket.pyx
      [4/8] Cythonizing zmq\devices\monitoredqueue.pyx
      [5/8] Cythonizing zmq\backend\cython\_device.pyx
      [6/8] Cythonizing zmq\backend\cython\_poll.pyx
      [7/8] Cythonizing zmq\backend\cython\_proxy_steerable.pyx
      [8/8] Cythonizing zmq\backend\cython\context.pyx
      C:\Users\nak179\AppData\Local\Temp\pip-build-env-27nrbhku\overlay\Lib\site-packages\setuptools\_distutils\dist.py:265: UserWarning: Unknown distribution option: 'cffi_modules'
        warnings.warn(msg)
      running egg_info
      writing pyzmq.egg-info\PKG-INFO
      writing dependency_links to pyzmq.egg-info\dependency_links.txt
      writing requirements to pyzmq.egg-info\requires.txt
      writing top-level names to pyzmq.egg-info\top_level.txt
      running configure
      error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.
note: This error originates from a subprocess, and is likely not a problem with pip.
</p>
</details>


I can use different envs, one for for training models using ivadomed with GPU support and and another one for using models with ADS without GPU. It is annoying, would be great if this issue can be resolved

**Reply by akraskovSMDB** [reply]
One more step, I installed  "Microsoft C++ Build Tools", now ADS installation breaks at a different point 
<details>
<summary>Details</summary>
<p>
Installing collected packages: wcwidth, snowballstemmer, pywin32, PyOpenGL, pure-eval, pickleshare, ipython-genutils, heapdict, executing, backcall, traitlets, tornado, toolz, tomli-w, tomli, sphinxcontrib-serializinghtml, sphinxcontrib-qthelp, sphinxcontrib-jsmath, sphinxcontrib-htmlhelp, sphinxcontrib-devhelp, sphinxcontrib-applehelp, rpds-py, qtpy, pyzmq, pydantic, prompt-toolkit, platformdirs, pkgutil-resolve-name, pint, parso, nest-asyncio, napari-plugin-engine, mypy-extensions, locket, Jinja2, in-n-out, imagesize, hsluv, fsspec, freetype-py, docutils, docstring-parser, decorator, debugpy, cloudpickle, cachey, babel, AxonDeepSeg, attrs, asttokens, alabaster, vispy, superqt, stack-data, sphinx, referencing, pyproject_hooks, psygnal, pooch, partd, matplotlib-inline, jupyter-core, jedi, comm, numpydoc, napari-svg, magicgui, jupyter-client, jsonschema-specifications, IPython, dask, build, app-model, npe2, jsonschema, ipykernel, qtconsole, napari-console, napari, napari-ADS
</p>
  Running setup.py develop for AxonDeepSeg
    error: subprocess-exited-with-error

    × python setup.py develop did not run successfully.
    │ exit code: 1
    ╰─> [92 lines of output]
        running develop
        running egg_info
        writing AxonDeepSeg.egg-info\PKG-INFO
        writing dependency_links to AxonDeepSeg.egg-info\dependency_links.txt
        writing entry points to AxonDeepSeg.egg-info\entry_points.txt
        writing top-level names to AxonDeepSeg.egg-info\top_level.txt
        reading manifest file 'AxonDeepSeg.egg-info\SOURCES.txt'
        adding license file 'LICENSE'
        writing manifest file 'AxonDeepSeg.egg-info\SOURCES.txt'
        running build_ext
        Creating c:\miniconda\envs\ads\lib\site-packages\AxonDeepSeg.egg-link (link to .)
        Adding AxonDeepSeg 4.1.0 to easy-install.pth file
        Installing axondeepseg-script.py script to c:\miniconda\envs\ads\Scripts
        Installing axondeepseg.exe script to c:\miniconda\envs\ads\Scripts
        Installing axondeepseg_morphometrics-script.py script to c:\miniconda\envs\ads\Scripts
        Installing axondeepseg_morphometrics.exe script to c:\miniconda\envs\ads\Scripts
        Installing axondeepseg_test-script.py script to c:\miniconda\envs\ads\Scripts
        Installing axondeepseg_test.exe script to c:\miniconda\envs\ads\Scripts
        Installing download_models-script.py script to c:\miniconda\envs\ads\Scripts
        Installing download_models.exe script to c:\miniconda\envs\ads\Scripts
        Installing download_tests-script.py script to c:\miniconda\envs\ads\Scripts
        Installing download_tests.exe script to c:\miniconda\envs\ads\Scripts

        Installed c:\miniconda\axondeepseg
        Traceback (most recent call last):
          File "\\?\C:\miniconda\envs\ads\Scripts\download_models-script.py", line 33, in <module>
            sys.exit(load_entry_point('AxonDeepSeg', 'console_scripts', 'download_models')())
          File "\\?\C:\miniconda\envs\ads\Scripts\download_models-script.py", line 25, in importlib_load_entry_point
            return next(matches).load()
          File "c:\miniconda\envs\ads\lib\importlib\metadata.py", line 77, in load
            module = import_module(match.group('module'))
          File "c:\miniconda\envs\ads\lib\importlib\__init__.py", line 127, in import_module
            return _bootstrap._gcd_import(name[level:], package, level)
          File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
          File "<frozen importlib._bootstrap>", line 991, in _find_and_load
          File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
          File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
          File "<frozen importlib._bootstrap_external>", line 843, in exec_module
          File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
          File "c:\miniconda\axondeepseg\AxonDeepSeg\download_model.py", line 1, in <module>
            from AxonDeepSeg.ads_utils import convert_path, download_data
          File "c:\miniconda\axondeepseg\AxonDeepSeg\ads_utils.py", line 17, in <module>
            import raven
        ModuleNotFoundError: No module named 'raven'
        c:\miniconda\envs\ads\lib\site-packages\setuptools\command\develop.py:40: EasyInstallDeprecationWarning: easy_install command is deprecated.
        !!

                ********************************************************************************
                Please avoid running ``setup.py`` and ``easy_install``.
                Instead, use pypa/build, pypa/installer or other
                standards-based tools.

                See https://github.com/pypa/setuptools/issues/917 for details.
                ********************************************************************************

        !!
          easy_install.initialize_options(self)
        c:\miniconda\envs\ads\lib\site-packages\setuptools\_distutils\cmd.py:66: SetuptoolsDeprecationWarning: setup.py install is deprecated.
        !!

                ********************************************************************************
                Please avoid running ``setup.py`` directly.
                Instead, use pypa/build, pypa/installer or other
                standards-based tools.

                See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.
                ********************************************************************************

        !!
          self.initialize_options()
        Traceback (most recent call last):
          File "<string>", line 2, in <module>
          File "<pip-setuptools-caller>", line 34, in <module>
          File "C:\miniconda\axondeepseg\setup.py", line 27, in <module>
            setup(
          File "c:\miniconda\envs\ads\lib\site-packages\setuptools\__init__.py", line 107, in setup
            return distutils.core.setup(**attrs)
          File "c:\miniconda\envs\ads\lib\site-packages\setuptools\_distutils\core.py", line 185, in setup
            return run_commands(dist)
          File "c:\miniconda\envs\ads\lib\site-packages\setuptools\_distutils\core.py", line 201, in run_commands
            dist.run_commands()
          File "c:\miniconda\envs\ads\lib\site-packages\setuptools\_distutils\dist.py", line 969, in run_commands
            self.run_command(cmd)
          File "c:\miniconda\envs\ads\lib\site-packages\setuptools\dist.py", line 1234, in run_command
            super().run_command(command)
          File "c:\miniconda\envs\ads\lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
            cmd_obj.run()
          File "C:\miniconda\axondeepseg\setup.py", line 22, in run
            check_call("download_models")
          File "c:\miniconda\envs\ads\lib\subprocess.py", line 364, in check_call
            raise CalledProcessError(retcode, cmd)
        subprocess.CalledProcessError: Command 'download_models' returned non-zero exit status 1.
        [end of output]

    note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× python setup.py develop did not run successfully.
│ exit code: 1
╰─> [92 lines of output]
    running develop
    running egg_info
    writing AxonDeepSeg.egg-info\PKG-INFO
    writing dependency_links to AxonDeepSeg.egg-info\dependency_links.txt
    writing entry points to AxonDeepSeg.egg-info\entry_points.txt
    writing top-level names to AxonDeepSeg.egg-info\top_level.txt
    reading manifest file 'AxonDeepSeg.egg-info\SOURCES.txt'
    adding license file 'LICENSE'
    writing manifest file 'AxonDeepSeg.egg-info\SOURCES.txt'
    running build_ext
    Creating c:\miniconda\envs\ads\lib\site-packages\AxonDeepSeg.egg-link (link to .)
    Adding AxonDeepSeg 4.1.0 to easy-install.pth file
    Installing axondeepseg-script.py script to c:\miniconda\envs\ads\Scripts
    Installing axondeepseg.exe script to c:\miniconda\envs\ads\Scripts
    Installing axondeepseg_morphometrics-script.py script to c:\miniconda\envs\ads\Scripts
    Installing axondeepseg_morphometrics.exe script to c:\miniconda\envs\ads\Scripts
    Installing axondeepseg_test-script.py script to c:\miniconda\envs\ads\Scripts
    Installing axondeepseg_test.exe script to c:\miniconda\envs\ads\Scripts
    Installing download_models-script.py script to c:\miniconda\envs\ads\Scripts
    Installing download_models.exe script to c:\miniconda\envs\ads\Scripts
    Installing download_tests-script.py script to c:\miniconda\envs\ads\Scripts
    Installing download_tests.exe script to c:\miniconda\envs\ads\Scripts

    Installed c:\miniconda\axondeepseg
    Traceback (most recent call last):
      File "\\?\C:\miniconda\envs\ads\Scripts\download_models-script.py", line 33, in <module>
        sys.exit(load_entry_point('AxonDeepSeg', 'console_scripts', 'download_models')())
      File "\\?\C:\miniconda\envs\ads\Scripts\download_models-script.py", line 25, in importlib_load_entry_point
        return next(matches).load()
      File "c:\miniconda\envs\ads\lib\importlib\metadata.py", line 77, in load
        module = import_module(match.group('module'))
      File "c:\miniconda\envs\ads\lib\importlib\__init__.py", line 127, in import_module
        return _bootstrap._gcd_import(name[level:], package, level)
      File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
      File "<frozen importlib._bootstrap>", line 991, in _find_and_load
      File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
      File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
      File "<frozen importlib._bootstrap_external>", line 843, in exec_module
      File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
      File "c:\miniconda\axondeepseg\AxonDeepSeg\download_model.py", line 1, in <module>
        from AxonDeepSeg.ads_utils import convert_path, download_data
      File "c:\miniconda\axondeepseg\AxonDeepSeg\ads_utils.py", line 17, in <module>
        import raven
    ModuleNotFoundError: No module named 'raven'
    c:\miniconda\envs\ads\lib\site-packages\setuptools\command\develop.py:40: EasyInstallDeprecationWarning: easy_install command is deprecated.
    !!

            ********************************************************************************
            Please avoid running ``setup.py`` and ``easy_install``.
            Instead, use pypa/build, pypa/installer or other
            standards-based tools.

            See https://github.com/pypa/setuptools/issues/917 for details.
            ********************************************************************************

    !!
      easy_install.initialize_options(self)
    c:\miniconda\envs\ads\lib\site-packages\setuptools\_distutils\cmd.py:66: SetuptoolsDeprecationWarning: setup.py install is deprecated.
    !!

            ********************************************************************************
            Please avoid running ``setup.py`` directly.
            Instead, use pypa/build, pypa/installer or other
            standards-based tools.

            See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.
            ********************************************************************************

    !!
      self.initialize_options()
    Traceback (most recent call last):
      File "<string>", line 2, in <module>
      File "<pip-setuptools-caller>", line 34, in <module>
      File "C:\miniconda\axondeepseg\setup.py", line 27, in <module>
        setup(
      File "c:\miniconda\envs\ads\lib\site-packages\setuptools\__init__.py", line 107, in setup
        return distutils.core.setup(**attrs)
      File "c:\miniconda\envs\ads\lib\site-packages\setuptools\_distutils\core.py", line 185, in setup
        return run_commands(dist)
      File "c:\miniconda\envs\ads\lib\site-packages\setuptools\_distutils\core.py", line 201, in run_commands
        dist.run_commands()
      File "c:\miniconda\envs\ads\lib\site-packages\setuptools\_distutils\dist.py", line 969, in run_commands
        self.run_command(cmd)
      File "c:\miniconda\envs\ads\lib\site-packages\setuptools\dist.py", line 1234, in run_command
        super().run_command(command)
      File "c:\miniconda\envs\ads\lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
        cmd_obj.run()
      File "C:\miniconda\axondeepseg\setup.py", line 22, in run
        check_call("download_models")
      File "c:\miniconda\envs\ads\lib\subprocess.py", line 364, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command 'download_models' returned non-zero exit status 1.
    [end of output]

note: This error originates from a subprocess, and is likely not a problem with pip.
</details>

**Reply by mathieuboudreau** [reply]
Looks like something went wrong with your ADS requirements installation, as the environment isn't seeing the raven package, but it’s listed in our environment file.

**Reply by akraskovSMDB** [reply]
I solved the problem 

installed pytroch which support GPU/CUDA and compatible with ivadomed 
```
conda install pytorch==1.11.0 torchvision==0.12.0 torchaudio==0.11.0 cudatoolkit=11.3 -c pytorch
```
tried to update this environment with ads environment.yml but coda is still trying to solve environment, maybe it will finish eventually

Instead, I tried to install packages from ads environment.yml by hand one by one, i.e. to solve it manually
The problem comes after 
```
pip install napari
````
It replaces pillow and imageio with newer versions (9.4.0 and 2.31.1) which are outside of requirements (pillow!=9.0.0, imageio<2.28.0) (and will produce errors in axosndeepseg). 

This one solves the problem
```
pip install "imageio<2.28.0" "pillow!=9.0.0" "napari"
```
which solved imageio to version 2.27 and pillow to 9.4.0 (which turned out to be ok ). 
Installed ads

And axosndeepseg_test worked and there is GPU/CUDA support

**Reply by Stoyan-I-A** [reply]
Oh yeah, the reason why we set imageio<2.28.0 is because of tests failing somewhere. More details in #740.

**Reply by mathieuboudreau** [reply]
Ah very interesing - that's great thanks for updating us! 

Odd that our Github Actions test suite didn't catch this, maybe because you were installing the packages manually whereas our test suite installs them via our environment.yml file all at once (which deals with a lot of the conflict management)

**Reply by akraskovSMDB** [reply]
It is a bit awkward way to solve the problem of not having GPU/CUDA support. It would be really nice if default ADS installation does it automatically.  Maybe it is more a question to ivadomed guys.

<details><summary>I played a bit and this environment file gives GPU/CUDA support </summary>

```
(ads2) C:\miniconda>sdiff ads_withpytorch1.yml axondeepseg/environment.yml
name: ads2                                                    | name: ads
channels:                                                       channels:
  - conda-forge                                                   - conda-forge
  - pytorch                                                   <
  - nvidia                                                    <
dependencies:                                                   dependencies:
  - python=3.8                                                    - python=3.8
  - pytorch=1.11.0                                            <
  - torchvision=0.12.0                                        <
  - torchaudio=0.11.0                                         <
  - cudatoolkit=11.3                                          <
  - nibabel 3.2.*                                                 - nibabel 3.2.*
  - h5py=2.10.0                                                   - h5py=2.10.0
  - numpy                                                         - numpy
  - scipy                                                         - scipy
  - scikit-learn=1.1.1                                            - scikit-learn=1.1.1
  - scikit-image=0.19.3                                           - scikit-image=0.19.3
  - tabulate                                                      - tabulate
  - pandas                                                        - pandas
  - matplotlib                                                    - matplotlib
  - mpld3                                                         - mpld3
  - tqdm                                                          - tqdm
  - requests                                                      - requests
  - pillow!=9.0.0                                                 - pillow!=9.0.0
  - imageio<2.28.0                                                - imageio<2.28.0
  - pytest                                                        - pytest
  - pytest-cov                                                    - pytest-cov
  - prettytable                                                   - prettytable
  - raven                                                         - raven
  - jupyter                                                       - jupyter
  - openpyxl                                                      - openpyxl
  - qtconsole<5.4.2                                               - qtconsole<5.4.2
  - pip                                                           - pip
  - pip:                                                          - pip:
    - napari                                                        - napari
    - protobuf==3.20.1                                              - protobuf==3.20.1
    - ivadomed~=2.9.8                                               - ivadomed~=2.9.8

```

</details>

I think the versions of pytorch related pkgs are max for the current ivadomed but it would be great if ivadomed can confirm this.
